### PR TITLE
*: fix some race condition

### PR DESCRIPTION
**Pull request by amyangfei**
### What problem does this PR solve? 
When I try to fix integration test coverage unstable issue, I found some race condition in code.

### What is changed and how it works?
Add some lock mechanism

### Check List 

Tests 

 - Integration test

### CODE REVIEW

**Comment by amyangfei** [comment]
PTAL @GregoryIan @csuzhangxc

File: relay/purger/purger.go [hunk 1]
```diff
@@ -54,6 +54,7 @@ const (
 
 // Purger purges relay log according to some strategies
 type Purger struct {
+	sync.RWMutex
```
Comment (csuzhangxc) [inline relay/purger/purger.go]: may use a named field is better (like `lock sync.RWMutex`), then outer code can't call methods from `RWMutex`. other similar usages can be updated later.

File: dm/master/server.go [hunk 2]
```diff
@@ -1599,14 +1599,15 @@ func (s *Server) allWorkerConfigs(ctx context.Context) (map[string]config.DBConf
...
+		errCh       = make(chan error, len(s.workerClients))
 		err         error
 	)
 	handErr := func(err2 error) {
 		workerMutex.Lock()
 		if err2 != nil {
 			log.Error(err2)
 		}
-		err = errors.Trace(err2)
+		errCh <- errors.Trace(err2)
```
[the remark this answers is not in the document]
Comment (amyangfei) [reply dm/master/server.go]: not needed. It is safe to write to a same channel from multiple goroutines.

**Comment by amyangfei** [comment]
I find another race condition when we processing with syncer jobs channel, although race condition will not detected. Sometimes we could get the following panic
```
panic: close of closed channel

goroutine 32 [running]:
github.com/pingcap/dm/syncer.(*Syncer).closeJobChans(0xc0004ad500)
    /home/jenkins/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/syncer/syncer.go:221 +0x10a
github.com/pingcap/dm/syncer.(*Syncer).stopSync(0xc0004ad500)
    /home/jenkins/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/syncer/syncer.go:1863 +0x8a
github.com/pingcap/dm/syncer.(*Syncer).Close(0xc0004ad500)
    /home/jenkins/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/syncer/syncer.go:1835 +0x120
github.com/pingcap/dm/dm/worker.(*SubTask).closeUnits(0xc00010efd0)
    /home/jenkins/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/dm/worker/subtask.go:265 +0x2a0
github.com/pingcap/dm/dm/worker.(*SubTask).Close(0xc00010efd0)
    /home/jenkins/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/dm/worker/subtask.go:349 +0x134
github.com/pingcap/dm/dm/worker.(*Worker).Close(0xc0000e69c0)
    /home/jenkins/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/dm/worker/worker.go:127 +0x1c6
github.com/pingcap/dm/dm/worker.(*Server).Close(0xc0001a6880)
    /home/jenkins/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/dm/worker/server.go:132 +0x171
```

PTAL @csuzhangxc

File: syncer/syncer.go [hunk 3]
```diff
@@ -217,9 +218,13 @@ func (s *Syncer) closeJobChans() {
 	if s.jobsClosed.Get() {
 		return
 	}
+
+	s.jobsChanLock.Lock()
```
Comment (csuzhangxc) [inline syncer/syncer.go]: whether needing to put this before `if s.jobsClosed.Get()` and put `s.jobsChanLock.Unlock()` after `s.jobsClosed.Set(true)`? and then `sync2.AtomicBool` can be changed to `bool`?

now, our concurrency control code is naive, we should refine this when doing the refactor.

Comment (amyangfei) [reply syncer/syncer.go]: oh yes, we should lock the whole `closeJobChans` function

File: relay/purger/purger.go [hunk 4]
```diff
@@ -113,7 +114,9 @@ func (p *Purger) run() {
 	defer ticker.Stop()
 
 	var ctx context.Context
+	p.lock.Lock()
```
Comment (GregoryIan) [inline relay/purger/purger.go]: still would override `p.cancel`, can we let only one goruotine running?

**Comment by GregoryIan** [comment]
Rest LGTM

File: relay/purger/purger.go (hunk 4 above)
Comment (amyangfei) [reply relay/purger/purger.go]: in current concurrent model a `Purger` will be started in one `DM-worker` only. So we don't need to lock the whole `run` function.
We'd better to refine concurrent usage later

Comment (GregoryIan) [reply relay/purger/purger.go]: if so, leave a `FIXME`

Comment (amyangfei) [reply relay/purger/purger.go]: added a note comment

**Comment by GregoryIan** [comment]
LGTM

**Pull request merged**